src/Features/CSharp/Portable/UseRecursivePatterns/AnalyzedNode.cs

```diff
@@ -0,0 +1,408 @@
+                            break;
+                        case True t:
+                            tests.Free();
+                            return t;
+                        case OrSequence seq:
+                            tests.RemoveAt(i);
+                            var testsToInsert = seq.Nodes;
+                            for (int j = 0, n = testsToInsert.Length; j < n; j++)
+                                tests.Insert(i + j, testsToInsert[j]);
+                            break;
+                    }
+                }
+                var result = tests.Count switch
+                {
+                    0 => False.Instance,
+                    1 => tests[0],
+                    _ => new OrSequence(tests.ToImmutable()),
+                };
+                tests.Free();
+                return result;
+            }
+        }
+
+        internal sealed class Pair : AnalyzedNode
+        {
+            public readonly Evaluation Input;
+            public readonly AnalyzedNode Pattern;
+
+            public Pair(Evaluation input, AnalyzedNode pattern)
+            {
+                Input = input;
+                Pattern = pattern;
+            }
+        }
+
+        internal abstract class Test : AnalyzedNode
+        {
+            public readonly Evaluation? Input;
+            protected Test(Evaluation? input) => Input = input;
+            public Test WithInput(Evaluation? newInput) => ReferenceEquals(newInput, Input) ? this : WithInputCore(newInput);
+            protected abstract Test WithInputCore(Evaluation? newInput);
+        }
+
+        internal sealed class Constant : Test
+        {
+            public readonly IOperation Value;
+            public Constant(Evaluation? input, IOperation value) : base(input)
+            {
+                Debug.Assert(value.ConstantValue.HasValue);
+                Value = value;
+            }
+
+            protected override Test WithInputCore(Evaluation? newInput) => new Constant(newInput, Value);
+        }
+
+        internal sealed class Relational : Test
+        {
+            public readonly BinaryOperatorKind OperatorKind;
+            public readonly IOperation Value;
+            public Relational(Evaluation? input, BinaryOperatorKind operatorKind, IOperation value) : base(input)
+            {
+                Debug.Assert(value.ConstantValue.HasValue);
+                OperatorKind = operatorKind;
+                Value = value;
+            }
+
+            protected override Test WithInputCore(Evaluation? newInput) => new Relational(newInput, OperatorKind, Value);
+        }
+
+        internal abstract class Evaluation : Test
+        {
+            // Record the syntax so we don't have to recreate the
+            // whole node if this happens to be rewritten as an expression.
+            public readonly SyntaxNode? Syntax;
+            protected Evaluation(Evaluation? input, SyntaxNode? syntax = null) : base(input) => Syntax = syntax;
```

Comment:
sorry, this is just getting too cute, please write the normal style of:
```suggestion
            protected Evaluation(Evaluation? input, SyntaxNode? syntax = null) : base(input)
                => Syntax = syntax;
```